File: julia/2021/day07.jl
using Statistics: median, mean

cost(d) = d * (1 + d) / 2

function main()
    input = parse.(Int, split(readline(joinpath(@__DIR__, "input/day07.txt")), ','))
    # input = parse.(Int, split("16,1,2,0,4,2,7,1,2,14", ','))

    m1 = median(input)
    m2 = mean(input)

    part1 = sum(Int, abs.(input .- m1))
    part2 = Int(minimum(sum(cost.(abs.(input .- mx))) for mx = round(m2 - 0.5):round(m2 + 0.5)))
    return part1, part2
end

println(main())

# part2
# cost(x, mx) = Sum((x - mx)^2) / 2 + Sum(abs(x - mx)) / 2
# Sum(2(x - mx)) / 2 + Sum(-1) / 2 <= dCostdmx(x) <= Sum(2(x - mx)) / 2 + Sum(+1) / 2
# dCostdmx(x) == 0
# -1 / 2 + Sum(x) / n <= mx <= +1 / 2 + Sum(x) / n

File: javascript/2021/src/day07.js
const { readFileSync } = require("fs");
const { mean, median, abs } = require("mathjs");

function pa_sum(d) {
  return (d * (1 + d)) / 2;
}

function day07(fileName = "./data/day07.txt") {
  let input = readFileSync(fileName, "utf8")
    .split(",")
    .map((i) => parseInt(i));
  //   input = [16, 1, 2, 0, 4, 2, 7, 1, 2, 14];
  m1 = Math.round(median(input));
  m2 = mean(input);
  m2_floor = Math.floor(m2);
  m2_ceil = Math.ceil(m2);

  const part1 = input.map((x) => Math.abs(x - m1)).reduce((a, b) => a + b, 0);
  const part2_floor = input
    .map((x) => pa_sum(Math.abs(x - m2_floor)))
    .reduce((a, b) => a + b, 0);
  const part2_ceil = input
    .map((x) => pa_sum(Math.abs(x - m2_ceil)))
    .reduce((a, b) => a + b, 0);

  const part2 = Math.min(part2_ceil, part2_floor);

  return [part1, part2];
}

module.exports = {
  day07,
};
